app/src/main/java/com/example/sharingang/items/ItemsViewModel.kt

```diff
@@ -1,114 +1,162 @@
-        _refreshing.value = true
-        viewModelScope.launch {
-            itemRepository.refreshItems()
-            // Since we're in a coroutine, need to use post instead
-            _refreshing.postValue(false)
-        }
-    }
-}
+package com.example.sharingang.items
+
+import android.util.Log
+import androidx.lifecycle.*
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+/**
+ * ItemsViewModel models the state of the fragment for viewing items
+ */
+@HiltViewModel
+class ItemsViewModel @Inject constructor(
+    private val itemRepository: ItemRepository
+) : ViewModel() {
+
+    init {
+        viewModelScope.launch {
+            itemRepository.refreshItems()
+        }
+    }
+
+    enum class OBSERVABLES{
+        ALL_ITEMS, SEARCH_RESULTS
+    }
+
+
+    private val _navigateToEditItem = MutableLiveData<Item?>()
+    val navigateToEditItem: LiveData<Item?>
+        get() = _navigateToEditItem
+
+    private val _focusedItem = MutableLiveData<Item?>()
+    val focusedItem: LiveData<Item?>
+        get() = _focusedItem
+
+    private val _viewingItem = MutableLiveData(false)
+    val viewingItem: LiveData<Boolean>
+        get() = _viewingItem
+
+    private val _refreshing = MutableLiveData(false)
+    val refreshing: LiveData<Boolean>
+        get() = _refreshing
+
+
+    private val _searchResults = MutableLiveData<List<Item>>(listOf())
+    val searchResults: LiveData<List<Item>>
+        get() = _searchResults
+
+    /**
+     * The last item created
+     */
+    val items: LiveData<List<Item>>
+        get() = itemRepository.items()
+
+    /**
+     * Add a new item.
+     *
+     * @param item the item to be added
+     */
+    fun addItem(item: Item) {
+        viewModelScope.launch(Dispatchers.IO) {
+            itemRepository.addItem(item)
+        }
+    }
+
+
+    fun clearSearchResults(){
+        _searchResults.value = listOf<Item>()
+    }
+
+    /**
+     * Searches through database, ignores searchName/
+     * categoryID if null/empty.
+     *
+     * @param searchName string searched for
+     * @param categoryID category searched for
+     */
+    fun searchItems(searchName: String, categoryID : Int){
+        viewModelScope.launch(Dispatchers.IO){
+            val allItemsSet = HashSet<Item>(itemRepository.getAllItems())
+            val categoryResults = HashSet<Item>(allItemsSet)
+            val nameResults = HashSet<Item>(allItemsSet)
+
+            if(categoryID != 0){
+                for(item in allItemsSet){
+                    if(item.category != categoryID){
+                        categoryResults.remove(item)
+                    }
+                }
+            }
+
+            if(searchName.isNotEmpty()){
+                for(item in allItemsSet){
+                    if(!item.title.toLowerCase().contains(searchName!!.toLowerCase())){
+                        nameResults.remove(item)

```

Comment:
I'm not really convinced by the algorithm. You're taking all the items and removing the ones that don't contain a certain string. Wouldn't it be better to start with an empty list and add items that have a certain string in their title?